docs/03-code-internals/24-customizing-topic-list.md

```diff
@@ -0,0 +1,90 @@
+
+## Adding, removing and rearranging columns
+
+Discourse default topic list on Desktop is rendered as a table. The [JS plugin API](https://meta.discourse.org/t/41281) can be used to customize the set of columns by adding, removing, replacing or reordering them. This surgical modification of columns means that your theme or plugin is making the smallest possible change, so that core changes are unlikely to conflict with it in future.
+
+The topic-list columns are represented by a 'DAG API' object, accessible via the `topic-list-columns` [Value Transformer](https://meta.discourse.org/t/349954). Each column is represented by a plain JS object with two keys:
+
+- `header`: An Ember Component to be rendered as the header cell for this column
+- `item`: An Ember Component to be rendered as the item cell for this column
+
+A list of arguments passed to header cells can be found [here](https://github.com/discourse/discourse/blob/b76c5406bd/app/assets/javascripts/discourse/app/components/topic-list/header.gjs#L6C5-L29C14), and a list of arguments passed to item cells can be found [here](https://github.com/discourse/discourse/blob/b76c5406bdd4a9277a0bfc85c54b92a78f6ce48a/app/assets/javascripts/discourse/app/components/topic-list/item.gjs#L384C11-L395C20). Core's default columns can be found [here](https://github.com/discourse/discourse/blob/b76c5406bdd4a9277a0bfc85c54b92a78f6ce48a/app/assets/javascripts/discourse/app/components/topic-list/list.gjs#L44-L88).
+
+This example adds, removes and repositions columns in the desktop topic list:
+
+```gjs
+// .../discourse/api-initializers/my-topic-list-customization.gjs
+
+import { apiInitializer } from "discourse/lib/api";
+
+const StaffHeaderCell = <template>
+  <th>Staff?</th>
+</template>;
+const StaffItemCell = <template>
+  <td>{{if @topic.creator.staff "✅"}}</td>
+</template>;
+
+export default apiInitializer("1.34", (api) => {
+  const discoveryService = api.container.lookup("service:discovery");
+
+  api.registerValueTransformer("topic-list-columns", ({ value: columns }) => {
+    // Remove the core column which shows poster avatars:
+    columns.delete("posters");
+
+    // Swap the "replies" and "views" columns:
+    columns.reposition("views", { before: "replies" });
+
+    // Lean on external autotracked state to make decisions:
+    if (discoveryService.category?.slug === "announcements") {
+      // Add a custom column:
+      columns.add("created-by-staff", {
+        header: StaffHeaderCell,
+        item: StaffItemCell,
+      });
+    }
+
+    return columns;
+  });
+});
+```
+
+## Introducing content via Plugin Outlets
+
+For the mobile view, and to add content inside existing cells of the desktop view, you can use Plugin Outlets. These can be identified and used in the same way as any other part of Discourse. Check out [the Plugin Outlet documentation](https://meta.discourse.org/t/using-plugin-outlet-connectors-from-a-theme-or-plugin/32727).
```

Comment:
> For the mobile view, and to add content inside existing cells of the desktop view

Te line above is a little confusing. Would it be more clear to just say "To add content inside cells you can use Plugin Outlets" ?

Reply:
The mobile view doesn't have "cells", so I think we do need to call it out specifically. Open to suggestions on improving the wording though!

Reply:
You can use Plugin Outlets to inject content into existing cells on desktop. You can also use them to inject content into multiple areas of the topic list on mbile.